src/JobCreator.cpp

```diff
@@ -0,0 +1,240 @@
+  try {
+    ModuleFactory = HDFWriterModuleRegistry::find(StreamSettings.Module);
+  } catch (std::exception const &E) {
+    throw std::runtime_error(
+        fmt::format("Error while getting '{}',  source: {}  what: {}",
+                    StreamSettings.Module, StreamSettings.Source, E.what()));
+  }
+
+  auto HDFWriterModule = ModuleFactory();
+  if (!HDFWriterModule) {
+    throw std::runtime_error(fmt::format(
+        "Can not create a HDFWriterModule for '{}'", StreamSettings.Module));
+  }
+
+  auto RootGroup = Task->hdfGroup();
+  try {
+    HDFWriterModule->parse_config(StreamSettings.ConfigStreamJson, "{}");
+  } catch (std::exception const &E) {
+    std::throw_with_nested(std::runtime_error(
+        fmt::format("Exception while HDFWriterModule::parse_config  module: {} "
+                    " source: {}  what: {}",
+                    StreamSettings.Module, StreamSettings.Source, E.what())));
+  }
+
+  auto StreamGroup = hdf5::node::get_group(
+      RootGroup, StreamSettings.StreamHDFInfoObj.HDFParentName);
+  HDFWriterModule->init_hdf({StreamGroup}, StreamSettings.Attributes);
+  HDFWriterModule->close();
+  HDFWriterModule.reset();
+}
+
+/// Helper to extract information about the provided streams.
+/// \param Logger Pointer to spdlog instance to be used for logging.
+static vector<StreamSettings>
+extractStreamInformationFromJson(std::unique_ptr<FileWriterTask> const &Task,
+                                 std::vector<StreamHDFInfo> &StreamHDFInfoList,
+                                 SharedLogger const &Logger) {
+  Logger->info("Command contains {} streams", StreamHDFInfoList.size());
+  std::vector<StreamSettings> StreamSettingsList;
+  for (auto &StreamHDFInfo : StreamHDFInfoList) {
+    try {
+      StreamSettingsList.push_back(
+          extractStreamInformationFromJsonForSource(StreamHDFInfo));
+      Logger->info("Adding stream: {}",
+                   StreamSettingsList.back().ConfigStreamJson);
+      setUpHdfStructure(StreamSettingsList.back(), Task);
+      StreamHDFInfo.InitialisedOk = true;
+    } catch (json::parse_error const &E) {
+      Logger->warn("Invalid json: {}", StreamHDFInfo.ConfigStream);
+      continue;
+    } catch (std::runtime_error const &E) {
+      Logger->warn("Exception while initialising writer module  what: {}  "
+                   "parent: {}  json: {}",
+                   E.what(), StreamHDFInfo.HDFParentName,
+                   StreamHDFInfo.ConfigStream);
+      continue;
+    } catch (...) {
+      Logger->error("Unknown error caught while trying to initialise stream  "
+                    "parent: {}  json: {}",
+                    StreamHDFInfo.HDFParentName, StreamHDFInfo.ConfigStream);
+    }
+  }
+  return StreamSettingsList;
+}
+
+std::unique_ptr<IStreamMaster> JobCreator::createFileWritingJob(
+    StartCommandInfo const &StartInfo,
+    std::shared_ptr<KafkaW::ProducerTopic> const &StatusProducer,
+    MainOpt &Settings) {
```

Comment:
I do not request a change as such here but I do have a question: Do we break the code if we turn `MainOpt &Settings` into `MainOpt Settings` (i.e. remove the reference)?

Reply:
No, I don't think so. It certainly builds at least